In mixed uniform/ParameterBlock syntax, Slang generates invalid Vulkan code

For this code:
```
import playground;

struct Test {
    float f_float;
};

uniform Test u_struct;
ParameterBlock<Test> pb_struct;
uniform Test u_struct_array[2];

RWStructuredBuffer<float> results;

[shader("compute")]
[numthreads(1, 1, 1)]
void compute_main(uint3 tid: SV_DispatchThreadID)
{
    if (any(tid != uint3(0)))
        return;

    results[0] = u_struct.f_float;
    results[1] = pb_struct.f_float;
    results[2] = u_struct_array[0].f_float;
    results[3] = u_struct_array[1].f_float;
}
```

Vulkan validation complains about:
```
[ERROR] (rhi) driver: error: -1520283006 - VUID-VkShaderModuleCreateInfo-pCode-08737:
vkCreateShaderModule(): pCreateInfo->pCode (spirv-val produced an error):
rules: A Block or BufferBlock cannot be nested within another Block or BufferBlock.
  %GlobalParams_natural = OpTypeStruct %Test_natural %_Array_natural_Test2
The Vulkan spec states: If pCode is a pointer to SPIR-V code, pCode must adhere to the validation rules described by the Validation Rules within a Module section of the SPIR-V Environment appendix (https://vulkan.lunarg.com/doc/view/1.4.309.0/windows/antora/spec/latest/chapters/shaders.html#VUID-VkShaderModuleCreateInfo-pCode-08737)
```

This blocks adding `ParameterBlock` as tested members of SlangPy's `test_shader_cursor`

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2107,6 +2107,96 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         }
     }
 
+    void legalizeStructBlocks()
+    {
+        // SPIRV does not allow using a struct with a block declaration as a field
+        // of another struct. Only top-level usage (e.g., global parameter blocks) should
+        // have the block decoration. If a struct is used both as a field and as a block,
+        // we must move the top-level usage to a wrapper struct, and move the block
+        // decoration to the wrapper struct.
+
+        HashSet<IRStructType*> embeddedBlockStructs;
+        for (auto globalInst : m_module->getGlobalInsts())
+        {
+            if (auto outerStruct = as<IRStructType>(globalInst))
+            {
+                for (auto field : outerStruct->getFields())
+                {
+                    if (auto innerStruct = as<IRStructType>(field->getFieldType()))
+                    {
+                        if (innerStruct->findDecorationImpl(kIROp_SPIRVBlockDecoration) ||
+                            innerStruct->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
+                        {
+                            embeddedBlockStructs.add(innerStruct);
+                        }
+                    }
+                }
+            }
+        }
+
+        for (auto globalInst : m_module->getGlobalInsts())
+        {
+            if (auto globalParam = as<IRGlobalParam>(globalInst))
+            {
+                auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType());
+                if (!ptrType) continue;
+                auto structType = as<IRStructType>(ptrType->getValueType());
+                if (!structType) continue;
+
+                if (embeddedBlockStructs.contains(structType))
+                {
+                    // Create a wrapper struct type
+                    IRBuilder builder(globalParam);
+                    builder.setInsertBefore(globalParam);
+
+                    auto wrapperStruct = builder.createStructType();
+                    auto key = builder.createStructKey();
+                    builder.createStructField(wrapperStruct, key, structType);
+
+                    // Copy the block decoration from the inner struct to the wrapper (do not remove yet)
+                    if (structType->findDecorationImpl(kIROp_SPIRVBlockDecoration))
+                    {
+                        builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBlockDecoration);
+                    }
+                    if (structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
+                    {
+                        builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBufferBlockDecoration);
+                    }
+
+                    // Update the global param's type to use the wrapper struct
+                    auto newPtrType = builder.getPtrType(ptrType->getOp(), wrapperStruct, ptrType->getAddressSpace());
+                    globalParam->setFullType(newPtrType);
+
+                    // Traverse all uses of the global param and insert a FieldAddress to access the inner struct
+                    traverseUses(
+                        globalParam,
+                        [&](IRUse* use)
+                        {
+                            builder.setInsertBefore(use->getUser());
+                            auto addr = builder.emitFieldAddress(
+                                builder.getPtrType(kIROp_PtrType, structType, ptrType->getAddressSpace()),
+                                globalParam,
+                                key);
+                            use->set(addr);
+                        });
+                }
+            }
+        }
+
+        // Remove block/buffer block decorations from all embedded block structs
+        for (auto structType : embeddedBlockStructs)
+        {
+            if (auto blockDecor = structType->findDecorationImpl(kIROp_SPIRVBlockDecoration))
+            {
+                blockDecor->removeAndDeallocate();
+            }
+            if (auto bufferBlockDecor = structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
+            {
+                bufferBlockDecor->removeAndDeallocate();
+            }
+        }
+    }
+
     void processModule()
     {
         determineSpirvVersion();
@@ -2195,6 +2285,10 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             m_sharedContext->m_targetProgram,
             m_module,
             bufferElementTypeLoweringOptions);
+
+        // Look for structs that are both used as fields and marked with Block
+        // decorations, and move the Block decoration to a wrapper struct.
+        legalizeStructBlocks();
 
         // Inline all pack/unpack storage type functions generated during buffer element
         // lowering pass.

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2116,6 +2116,7 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         // decoration to the wrapper struct.
 
         HashSet<IRStructType*> embeddedBlockStructs;
+        List<IRGlobalParam*> structGlobalParams;
         for (auto globalInst : m_module->getGlobalInsts())
         {
             if (auto outerStruct = as<IRStructType>(globalInst))
@@ -2132,66 +2133,69 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
                     }
                 }
             }
-        }
-
-        for (auto globalInst : m_module->getGlobalInsts())
-        {
-            if (auto globalParam = as<IRGlobalParam>(globalInst))
+            else if (auto globalParam = as<IRGlobalParam>(globalInst))
             {
-                auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType());
-                if (!ptrType)
-                    continue;
-                auto structType = as<IRStructType>(ptrType->getValueType());
-                if (!structType)
-                    continue;
-
-                if (embeddedBlockStructs.contains(structType))
+                if (auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType()))
                 {
-                    // Create a wrapper struct type
-                    IRBuilder builder(globalParam);
-                    builder.setInsertBefore(globalParam);
-
-                    auto wrapperStruct = builder.createStructType();
-                    auto key = builder.createStructKey();
-                    builder.createStructField(wrapperStruct, key, structType);
-
-                    // Copy the block decoration from the inner struct to the wrapper
-                    if (structType->findDecorationImpl(kIROp_SPIRVBlockDecoration))
+                    if (auto structType = as<IRStructType>(ptrType->getValueType()))
                     {
-                        builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBlockDecoration);
+                        structGlobalParams.add(globalParam);
                     }
-                    if (structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
-                    {
-                        builder.addDecorationIfNotExist(
-                            wrapperStruct,
-                            kIROp_SPIRVBufferBlockDecoration);
-                    }
-
-                    // Update the global param's type to use the wrapper struct
-                    auto newPtrType = builder.getPtrType(
-                        ptrType->getOp(),
-                        wrapperStruct,
-                        ptrType->getAddressSpace());
-                    globalParam->setFullType(newPtrType);
-
-                    // Traverse all uses of the global param and insert a FieldAddress to access the
-                    // inner struct
-                    traverseUses(
-                        globalParam,
-                        [&](IRUse* use)
-                        {
-                            builder.setInsertBefore(use->getUser());
-                            auto addr = builder.emitFieldAddress(
-                                builder.getPtrType(
-                                    kIROp_PtrType,
-                                    structType,
-                                    ptrType->getAddressSpace()),
-                                globalParam,
-                                key);
-                            use->set(addr);
-                        });
                 }
             }
+        }
+
+        for (auto globalParam : structGlobalParams)
+        {
+            auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType());
+            auto structType = as<IRStructType>(ptrType->getValueType());
+
+            if (!embeddedBlockStructs.contains(structType))
+                continue;
+
+            // Create a wrapper struct type
+            IRBuilder builder(globalParam);
+            builder.setInsertBefore(globalParam);
+
+            auto wrapperStruct = builder.createStructType();
+            auto key = builder.createStructKey();
+            builder.createStructField(wrapperStruct, key, structType);
+
+            // Copy the block decoration from the inner struct to the wrapper
+            if (structType->findDecorationImpl(kIROp_SPIRVBlockDecoration))
+            {
+                builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBlockDecoration);
+            }
+            if (structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
+            {
+                builder.addDecorationIfNotExist(
+                    wrapperStruct,
+                    kIROp_SPIRVBufferBlockDecoration);
+            }
+
+            // Update the global param's type to use the wrapper struct
+            auto newPtrType = builder.getPtrType(
+                ptrType->getOp(),
+                wrapperStruct,
+                ptrType->getAddressSpace());
+            globalParam->setFullType(newPtrType);
+
+            // Traverse all uses of the global param and insert a FieldAddress to access the
+            // inner struct
+            traverseUses(
+                globalParam,
+                [&](IRUse* use)
+                {
+                    builder.setInsertBefore(use->getUser());
+                    auto addr = builder.emitFieldAddress(
+                        builder.getPtrType(
+                            kIROp_PtrType,
+                            structType,
+                            ptrType->getAddressSpace()),
+                        globalParam,
+                        key);
+                    use->set(addr);
+                });
         }
 
         // Remove block/buffer block decorations from all embedded block structs

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2172,7 +2172,8 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             }
 
             // Update the global param's type to use the wrapper struct
-            auto newPtrType = builder.getPtrType(ptrType->getOp(), wrapperStruct, ptrType->getAddressSpace());
+            auto newPtrType =
+                builder.getPtrType(ptrType->getOp(), wrapperStruct, ptrType->getAddressSpace());
             globalParam->setFullType(newPtrType);
 
             // Traverse all uses of the global param and insert a FieldAddress to access the

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2137,7 +2137,7 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             {
                 if (auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType()))
                 {
-                    if (auto structType = as<IRStructType>(ptrType->getValueType()))
+                    if (as<IRStructType>(ptrType->getValueType()))
                     {
                         structGlobalParams.add(globalParam);
                     }

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2168,16 +2168,11 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             }
             if (structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
             {
-                builder.addDecorationIfNotExist(
-                    wrapperStruct,
-                    kIROp_SPIRVBufferBlockDecoration);
+                builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBufferBlockDecoration);
             }
 
             // Update the global param's type to use the wrapper struct
-            auto newPtrType = builder.getPtrType(
-                ptrType->getOp(),
-                wrapperStruct,
-                ptrType->getAddressSpace());
+            auto newPtrType = builder.getPtrType(ptrType->getOp(), wrapperStruct, ptrType->getAddressSpace());
             globalParam->setFullType(newPtrType);
 
             // Traverse all uses of the global param and insert a FieldAddress to access the
@@ -2188,10 +2183,7 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
                 {
                     builder.setInsertBefore(use->getUser());
                     auto addr = builder.emitFieldAddress(
-                        builder.getPtrType(
-                            kIROp_PtrType,
-                            structType,
-                            ptrType->getAddressSpace()),
+                        builder.getPtrType(kIROp_PtrType, structType, ptrType->getAddressSpace()),
                         globalParam,
                         key);
                     use->set(addr);

--- a/source/slang/slang-ir-spirv-legalize.cpp
+++ b/source/slang/slang-ir-spirv-legalize.cpp
@@ -2139,9 +2139,11 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             if (auto globalParam = as<IRGlobalParam>(globalInst))
             {
                 auto ptrType = as<IRPtrTypeBase>(globalParam->getDataType());
-                if (!ptrType) continue;
+                if (!ptrType)
+                    continue;
                 auto structType = as<IRStructType>(ptrType->getValueType());
-                if (!structType) continue;
+                if (!structType)
+                    continue;
 
                 if (embeddedBlockStructs.contains(structType))
                 {
@@ -2153,28 +2155,37 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
                     auto key = builder.createStructKey();
                     builder.createStructField(wrapperStruct, key, structType);
 
-                    // Copy the block decoration from the inner struct to the wrapper (do not remove yet)
+                    // Copy the block decoration from the inner struct to the wrapper
                     if (structType->findDecorationImpl(kIROp_SPIRVBlockDecoration))
                     {
                         builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBlockDecoration);
                     }
                     if (structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
                     {
-                        builder.addDecorationIfNotExist(wrapperStruct, kIROp_SPIRVBufferBlockDecoration);
+                        builder.addDecorationIfNotExist(
+                            wrapperStruct,
+                            kIROp_SPIRVBufferBlockDecoration);
                     }
 
                     // Update the global param's type to use the wrapper struct
-                    auto newPtrType = builder.getPtrType(ptrType->getOp(), wrapperStruct, ptrType->getAddressSpace());
+                    auto newPtrType = builder.getPtrType(
+                        ptrType->getOp(),
+                        wrapperStruct,
+                        ptrType->getAddressSpace());
                     globalParam->setFullType(newPtrType);
 
-                    // Traverse all uses of the global param and insert a FieldAddress to access the inner struct
+                    // Traverse all uses of the global param and insert a FieldAddress to access the
+                    // inner struct
                     traverseUses(
                         globalParam,
                         [&](IRUse* use)
                         {
                             builder.setInsertBefore(use->getUser());
                             auto addr = builder.emitFieldAddress(
-                                builder.getPtrType(kIROp_PtrType, structType, ptrType->getAddressSpace()),
+                                builder.getPtrType(
+                                    kIROp_PtrType,
+                                    structType,
+                                    ptrType->getAddressSpace()),
                                 globalParam,
                                 key);
                             use->set(addr);
@@ -2190,7 +2201,8 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
             {
                 blockDecor->removeAndDeallocate();
             }
-            if (auto bufferBlockDecor = structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
+            if (auto bufferBlockDecor =
+                    structType->findDecorationImpl(kIROp_SPIRVBufferBlockDecoration))
             {
                 bufferBlockDecor->removeAndDeallocate();
             }

--- a/tests/bugs/gh-7431.slang
+++ b/tests/bugs/gh-7431.slang
@@ -1,0 +1,37 @@
+//TEST(compute):COMPARE_COMPUTE(filecheck-buffer=CHECKOUT):-vk -compute -output-using-type
+
+// CHECKOUT:      10
+// CHECKOUT-NEXT: 20
+// CHECKOUT-NEXT: 30
+// CHECKOUT-NEXT: 40
+
+struct Test {
+    int f_int;
+};
+
+ParameterBlock<Test> pb_struct;
+uniform Test u_struct;
+uniform Test u_struct_array[2];
+
+RWStructuredBuffer<int> results;
+
+//TEST_INPUT: set pb_struct = new Test{10};
+//TEST_INPUT: uniform(data=[20 0 0 0]):name=u_struct.f_int
+//TEST_INPUT: uniform(data=[30 0 0 0]):name=u_struct_array[0].f_int
+//TEST_INPUT: uniform(data=[40 0 0 0]):name=u_struct_array[1].f_int
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=results
+
+[shader("compute")]
+[numthreads(1, 1, 1)]
+void computeMain(uint3 tid: SV_DispatchThreadID)
+{
+    if (any(tid != uint3(0)))
+        return;
+
+    results[0] = pb_struct.f_int;
+    results[1] = u_struct.f_int;
+    results[2] = u_struct_array[0].f_int;
+    results[3] = u_struct_array[1].f_int;
+}
+

